drivers/serial/uart_nrfx_uarte.c

```diff
@@ -1611,10 +1618,33 @@ static int uarte_nrfx_fifo_read(const struct device *dev,
 		/* Clear the interrupt */
 		nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_ENDRX);
 
-		/* Receive a character */
-		rx_data[num_rx++] = (uint8_t)data->rx_data;
+		uint32_t rx_amount = nrf_uarte_rx_amount_get(uarte);
+
+		num_rx = MIN(size, rx_amount);
 
-		nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
+		memcpy(rx_data,
+		       data->int_driven->rx_buffer + data->int_driven->rx_index,
+		       num_rx);
+		data->int_driven->rx_index = 0;
+
+		if (size > num_rx) {
+			/* Attempt reading the remaining bytes in a single transfer. */
+			nrf_uarte_rx_buffer_set(uarte,
+						data->int_driven->rx_buffer,
+						MIN(data->int_driven->rx_buff_size,
+						    size - num_rx));
+			nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
```

Comment:
The problem with this is that a new RX interrupt will not be generated until this buffer is completely filled up and `uart_irq_rx_ready()` will be returning 0 until that happens. And this is not the way the interrupt driven API is supposed to work. https://github.com/zephyrproject-rtos/zephyr/blob/8b4407ab7cfb8319c3bf57cb01cbcb8cf9b7eb46/include/zephyr/drivers/uart.h#L1008-L1011
When a user calls for example `uart_fifo_read(dev, buf, 8)`, it does not necessarily mean that 8 characters is actually needed. It only means that at most 8 characters can be written to the passed buffer. See also:
https://github.com/zephyrproject-rtos/zephyr/blob/8b4407ab7cfb8319c3bf57cb01cbcb8cf9b7eb46/include/zephyr/drivers/uart.h#L772-L775

If you want to make use of DMA, use the asynchronous API instead (that's actually the reason why this API was introduced).

Reply:
thanks for the explanation. So what you're saying is that the zephyr interrupt-driven uart API will only ever allow 1 byte to be read at a time, and no DMA usage, correct? As in, this is not a limitation of our implementation or HW, just a limitation of the API itself.